Add information about forecasted points change at start of combat. Closes #1365

bopke opened the pull request:
In depths of woodland realms unchartered, the tracks too easily criss-cross back and over themselves, losing all meaning under shades of bush and fern,
and only those who studiously read those glyphs of hoof and paw, may then follow them down wilder, freer, paths.
For some time you have yearned to speak this strange and peculiar tongue of beast, blood and soil,
and a beying echo from far away in the greenery, promises to teach you.
Tracking the Hunter's Trail promises...

When great warriors clash, the suspense and anticipation, in the heat of battle, can be turned down by a single thought - what are the repercussions of this act? Are the ramifications really worth it? This one change will end the misery and suffering of knight pondering what will happen, when the battle dust settles. Let him know upfront, when he engages in the fight, how many points will be awarded to him, apart from glory of the battle he won.


<img width="1701" height="511" alt="image" src="https://github.com/user-attachments/assets/8761c96f-f798-4bd9-9bd6-d179b00a2d93" />

Resolves #1365

bopke committed:
Add information about forecasted points change at start of combat

bopke commented:
This PR is made as a draft because in 7 days it'll be 5th birthday of the issue. Let's merge it then. Until this day comes, let's iterate on the issue so it'll be ready to be merged in time for the great celebration.

bopke commented on plugin/src/main/java/net/dzikoysk/funnyguilds/damage/DamageState.java:
```diff
@@ -50,6 +50,7 @@ public double getTotalDamage() {
     }
 
     public double getTotalDamage(User user) {
+        this.update();
```
I believe that every DamageState public facing method should call update first, to avoid serving stale data. This seems like a good change that will keep returned data correct, as we shouldn't operate on expired damage data.

I only changed it for getTotalDamage(User) as this was needed for my change to work. If you think that this is correct, I will add this update to all other methods. If this is not right, I would need to create other mean of tracking damage times but I don't think this is a good solution to the issue at hand.

Kamilkime replied:
I guess it all depends on how often this would be called. I think we can try adding that everywhere, but in the long run we should probably have a better system, which doesn't require constant updates and re-calculation of everything

bopke commented on plugin/src/main/java/net/dzikoysk/funnyguilds/damage/DamageState.java:
```diff
@@ -111,7 +112,7 @@ public void update() {
         DamageTracking trackingConfig = FunnyGuilds.getInstance().getPluginConfiguration().damageTracking;
 
         // Remove expired
-        this.damageHistory.removeIf(damage -> this.damageHistory.size() > 1 && damage.isExpired(trackingConfig.expireTime));
+        this.damageHistory.removeIf(damage -> damage.isExpired(trackingConfig.expireTime));
```
I removed that so we don't keep singular stale data even after update() being called. It looks like it shouldn't break anything else, but I have only did some simple testing of other functionalities that depend on DamageState, so it still can have some additional ramifications that are beyond my quick testing. I believe that this should be the correct way for the history to operate.

Kamilkime replied:
You should also take a look at `PlayerDeath#handleAssists`. We create a copy of the damage map there, but still call `getTotalDamage` on the original state. So with your changes - there could be an update of the original state after the copy was already created. I'm curious if that could maybe break something now

P3ridot commented:
changes where

dzikoysk committed:
Merge branch '5.x' into Add-information-about-points-change-in-combat

dzikoysk committed:
Revert DamageState changes from this PR

The forecast feature does not need to call update() on every read or to
prune the last surviving damage entry — coupling the cooldown to
damageTracking.expireTime side-affects getLastDamage(), isInCombat(), and
the considerLastAttackerAsKiller path. The next commit replaces the
gating with a dedicated cooldown that doesn't touch DamageState.

dzikoysk committed:
Replace previousDamage gate with a dedicated cooldown

The forecast title now fires at most once per (attacker, victim) pair per
combat-prediction-interval (default 30s) instead of being driven by the
side effect of damageHistory's expiry. Lets the spam interval be tuned
independently of damageTracking.expireTime.

Also tidy: align placeholders with rankDeathMessage's
{ATTACKER-CHANGE-FORMATTED}/{VICTIM-CHANGE-FORMATTED} convention; sort
the new RankSystem import; drop the round-trip negation/abs in favor of
passing the signed change straight to formatChange().

dzikoysk committed:
Expose predicted gain/loss as placeholders in /info

P3ridot suggested in #1365 that the same data should be available under
/info. Sender-relative {POINTS-GAIN}, {POINTS-LOSS} and their formatted
variants are now resolved per-receiver in PlayerInfoCommand and listed
in the playerInfoList / playerRightClickInfo comments.

dzikoysk committed:
Move forecast send before assist gates and bound the cooldown map

Two issues from copilot's review:
- The previous version was unreachable when assistEnable was false or in
  WorldGuard non-assists regions, even though the actual kill/points
  change still happens there. Move the forecast call before those gates.
- The plain ConcurrentHashMap held cooldown entries forever. Switch to a
  Caffeine cache with a per-entry Expiry that reads the configured
  interval at write time, so memory stays bounded by active pairs and a
  reload picks up new intervals on next entry.

dzikoysk committed:
Extract RankFormatter.formatPointsChange shared helper

The same NumberRange.inRangeToString(...).replace("{CHANGE}", abs)
pattern was now in three places: PlayerDeath inline, EntityDamage, and
PlayerInfoCommand. Centralize it so the formatting rule for kill-point
changes lives in one place.

dzikoysk committed:
Validate combat-prediction-interval as @PositiveOrZero

A negative duration would silently disable the cooldown by making
lastFire.plus(interval).isAfter(now) always return false. Reject it at
config load instead.

dzikoysk committed:
Reword combat prediction defaults to avoid double-negation

{VICTIM-CHANGE-FORMATTED} expands to a signed value like "-10", so the
old default "Stracisz {VICTIM-CHANGE-FORMATTED} punktów" rendered as
"Stracisz -10 punktów" — semantically backwards. Use a neutral
"Punkty z walki: {…-CHANGE-FORMATTED}" on both sides; the color and
sign in the formatted token already carry the gain/loss meaning.
